KAFKA-15859: Make RemoteListOffsets call an async operation

kamalcph opened the pull request:
This is the part-2 of the [KIP-1075](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1075%3A+Introduce+delayed+remote+list+offsets+purgatory+to+make+LIST_OFFSETS+async)

To find the offset for a given timestamp, ListOffsets API is used by the client. When the topic is enabled with remote storage, then we have to fetch the remote indexes such as offset-index and time-index to serve the query. Also, the ListOffsets request can contain the query for multiple topics/partitions.

The time taken to read the indexes from remote storage is non-deterministic and the query is handled by the request-handler threads. If there are multiple LIST_OFFSETS queries and most of the request-handler threads are busy in reading the data from remote storage, then the other high-priority requests such as FETCH and PRODUCE might starve and be queued. This can lead to higher latency in producing/consuming messages.

In this patch, we have introduced a delayed operation for remote list-offsets call. If the timestamp need to be searched in the remote-storage, then the request-handler threads will pass-on the request to the remote-log-reader threads. And, the request gets handled in asynchronous fashion.

Covered the patch with unit and integration tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kamalcph committed:
KAFKA-15859: Make RemoteListOffsets call an async operation

To find the offset for a given timestamp, ListOffsets API is used by the client. When the topic is enabled with remote storage, then we have to fetch the remote indexes such as offset-index and time-index to serve the query. Also, the ListOffsets request can contain the query for multiple topics/partitions.

The time taken to read the indexes from remote storage is non-deterministic and the query is handled by the request-handler threads. If there are multiple LIST_OFFSETS queries and most of the request-handler threads are busy in reading the data from remote storage, then the other high-priority requests such as PRODUCE and FETCH won't be handled and in worst-case can be dropped.

In this patch, we have introduced a delayed operation for remote list-offsets call. If the offsets need to be searched in the remote-storage, then the request-handler threads will pass-on the request to the remote-log-reader threads. And, the request gets handled in asynchronous fashion.

kamalcph commented:
@chia7712 @showuon @satishd 

Call for review. PTAL.

kamalcph commented:
Test failures are unrelated.

kamalcph commented:
Need some clarifications on defining the timeout for the delayed remote LIST_OFFSETS request. Currently, the LIST_OFFSETS API is invoked by:

1. [KafkaConsumer#offsetForTimes(Map<TopicPartition, Long> timestampsToSearch)](https://javadoc.io/doc/org.apache.kafka/kafka-clients/latest/org/apache/kafka/clients/consumer/KafkaConsumer.html#offsetsForTimes-java.util.Map-) 
2. [KafkaConsumer#offsetForTimes(Map<TopicPartition, Long> timestampsToSearch, Duration timeout)](https://javadoc.io/doc/org.apache.kafka/kafka-clients/latest/org/apache/kafka/clients/consumer/KafkaConsumer.html#offsetsForTimes-java.util.Map-java.time.Duration-) and
3. [Admin#listOffsets(Map<TopicPartition, OffsetSpec> topicPartitionOffsets)](https://javadoc.io/static/org.apache.kafka/kafka-clients/3.7.1/org/apache/kafka/clients/admin/Admin.html#listOffsets-java.util.Map-)

Both KafkaConsumer and Admin defines the default api-timeout to 60 seconds and default request-timeout to 30 seconds. There is no upper timeout boundary for handling the LIST_OFFSETS request on the server side and the client does not convey any timeout in the [ListOffsetsRequest.json](https://github.com/apache/kafka/blob/trunk/clients/src/main/resources/common/message/ListOffsetsRequest.json) request. 

**Current behavior**

`KafkaConsumer` sends one or more requests internally depends on the configured request-timeout and default-api-timeout. (eg) If the consumer is configured with default-api-timeout as 5 mins and request-timeout as 30 seconds. And, the server takes 50 seconds to process the LIST_OFFSETS request, then the consumer internally sends 10 LIST_OFFSETS requests before failing with TimeoutException after 5 minutes. 

`Admin` sends only one request and wait for upto default-api-timeout. (eg) If the admin is configured with default-api-timeout as 5 mins and request-timeout as 30 seconds. And, the server takes 50 seconds to process the LIST_OFFSETS request, then the admin sends only one LIST_OFFSETS request, then receives the request from server after 50 seconds. 


**Options**

1. If we add a new timeout field in the [ListOffsetsRequest](https://github.com/apache/kafka/blob/trunk/clients/src/main/resources/common/message/ListOffsetsRequest.json), then only new clients can able to specify the timeout. What timeout should we set for the requests from older clients?
2. If we propose a new dynamic `remote.list.offsets.timeout.ms` timeout on the server end, then only the operator of the cluster can be able to change the timeout. Most of the clients won't be having the cluster level permissions to update the timeout value on the server.


cc @satishd @showuon @chia7712

kamalcph commented:
Since tiered storage is not production ready, reusing the server request-timeout as the timeout for delayed remote LIST_OFFSETS request. Let me know whether we need a KIP to proceed forward with this PR. 

Is it possible to merge this in 3.9 branch? Thanks!

We are planning to propose a KIP once I get clarity on the where to define the timeout for the LIST_OFFSETS request.

showuon commented:
@kamalcph , I'd like to make sure this PR can be reviewed before [KIP-1075](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1075%3A+Introduce+delayed+remote+list+offsets+purgatory+to+make+LIST_OFFSETS+async) get approved. Is that right?

kamalcph commented:
> @kamalcph , I'd like to make sure this PR can be reviewed before [KIP-1075](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1075%3A+Introduce+delayed+remote+list+offsets+purgatory+to+make+LIST_OFFSETS+async) get approved. Is that right?

yes, this PR can be reviewed. There are no public API changes made in this PR. To define the timeout for delayed remote list offsets operation, reused the server request timeout since the tiered storage is not production ready. If it is not acceptable, then we may have to wait for the KIP-1075 approval.

showuon commented:
Well, I'll review KIP-1075 first then.

kamalcph commented:
@showuon @clolov 

Is it possible to give this PR an early review to keep it in good shape? Thanks!

showuon reviewed (commented):
Had a quick pass, I understand more about what we're trying to achieve here. High-level comment: I saw there are some new added files are in scala, could we re-write into java? You don't have to do that now, please wait until KIP accepted. Let's discuss more detail in KIP discussion. Thanks.

chia7712 commented:
@kamalcph please fix conflicts, thanks :)

kamalcph committed:
Emit DelayedRemoteListOffsetsMetrics

kamalcph commented:
@showuon @chia7712 @satishd @clolov 

The diff is ready for review. PTAL. Thanks!

showuon commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -1601,21 +1601,25 @@ class Partition(val topicPartition: TopicPartition,
         getOffsetByTimestamp
       case _ =>
-        getOffsetByTimestamp.filter(timestampAndOffset => timestampAndOffset.offset < lastFetchableOffset)
-          .orElse(maybeOffsetsError.map(e => throw e))
+        val offsetResultHolder = getOffsetByTimestamp
+        offsetResultHolder.maybeOffsetsError = maybeOffsetsError
+        offsetResultHolder.lastFetchableOffset = Some(lastFetchableOffset)
+        offsetResultHolder
     }
```
Could you help me understand this change please?

kamalcph replied:
Once the `getOffsetByTimestamp` returns an offset. If the offset is higher than the `lastFetchableOffset` (either LSO or HW), then we should throw retriable `OFFSET_NOT_AVAILABLE` error. 

With remote storage, the `getOffsetByTimestamp` becomes async, so we are storing the  `maybeOffsetsError` and  `lastFetchableOffset` in the offsetResultHolder and use them in the DelayedRemoteListOffsets once we get the response from remote.

showuon replied:
OK, I see.

showuon reviewed (commented):
Will continue the review later.

showuon commented on core/src/test/java/kafka/log/remote/RemoteLogOffsetReaderTest.java:
```diff
@@ -0,0 +1,196 @@
+        assertThrows(RejectedExecutionException.class, () ->
+                holderList.add(rlm.asyncOffsetRead(topicPartition, time.milliseconds(), 0L, cache, Option::empty)));
+
+        holderList.get(2).jobFuture().cancel(false);
```
Why does the job cancel doesn't complete the task?

kamalcph replied:
The reason for maintaining 2 futures: `jobFuture` and `taskFuture` is to trigger the delayed operation completion (`delayedRemoteListOffsetsPurgatory#checkAndComplete(key)`) in the same remote-log-reader thread. When all the partitions results are received, then the delayedOperation gets completed.

There is no relationship between the `jobFuture` and `taskFuture`. We call `jobFuture#cancel` only when the delayedOperation gets expired. So, either completing the `taskFuture` or not won't have any effect as the `taskFuture` result won't be used.

showuon replied:
I see. Thanks.

showuon commented on storage/src/test/java/org/apache/kafka/tiered/storage/actions/ExpectListOffsetsAction.java:
```diff
@@ -19,33 +19,50 @@
     @Override
     public void doExecute(TieredStorageTestContext context) throws InterruptedException, ExecutionException {
+        if (expected.epoch == Integer.MAX_VALUE) {
+            // The leaderEpoch for the partition gets bumped/incremented when the partition gets reassigned (or)
+            // a new leader gets elected. But, the leaderEpoch value is non-deterministic between the ZK and Kraft mode.
+            // So, reading the leader-epoch value from the broker's LeaderEpochFileCache for the given offset.
+            Optional<LeaderEpochFileCache> leaderEpochFileCache = context.leaderEpochFileCache(brokerId, partition);
+            assertTrue(leaderEpochFileCache.isPresent());
+            leaderEpochFileCache.get().epochForOffset(expected.startOffset).ifPresent(epoch ->
+                expected = new EpochEntry(epoch, expected.startOffset)
```
Could you explain more why the leaderEpoch is non-deterministic? This works though, it makes the testSpec is not readable when verifying listOffset.

kamalcph replied:
I tried running the ListOffsetsTest for both ZK and Kraft modes:

1. In ZK mode, the leader-epoch gets bumped during reassignment (0 -> 1 -> 2) and leader-election (2 -> 3)
2. In Kraft mode, the leader-epoch gets bumped only for leader-election (0 -> 1) and not for reassignment

So, the test fails for one mode and succeeds in another due to non-deterministic leader-epoch bump between Kraft and ZK modes. Since ZK mode is deprecated now, I'll run this test only for Kraft mode.

showuon commented on storage/src/test/java/org/apache/kafka/tiered/storage/integration/ListOffsetsTest.java:
```diff
@@ -0,0 +1,111 @@
+                .deleteRecords(topicA, p0, 4L)
+                .expectListOffsets(broker1, topicA, p0, OffsetSpec.earliest(), new EpochEntry(Integer.MAX_VALUE, 4))
+                .expectListOffsets(broker1, topicA, p0, OffsetSpec.earliestLocal(), new EpochEntry(Integer.MAX_VALUE, 4))
+                .expectListOffsets(broker1, topicA, p0, OffsetSpec.latestTiered(), new EpochEntry(-1, 3))
```
nit: Let's use `RecordBatch.NO_PARTITION_LEADER_EPOCH` instead of `-1`.

showuon commented on storage/src/test/java/org/apache/kafka/tiered/storage/integration/ListOffsetsTest.java:
```diff
@@ -0,0 +1,111 @@
+                        new KeyValueSpec("k3", "v3", timestamp + 2))
+
+                // switch leader and send more records to partition 0 and expect the second segment to be offloaded.
+                // On reassignment, leader-epoch gets bumped from 0 -> 1 -> 2
+                .reassignReplica(topicA, p0, Arrays.asList(broker1, broker0))
+                // On leader election, leader-epoch gets bumped from 2 -> 3
```
These 2 comments didn't reflect in the below assertion.

showuon commented on core/src/main/scala/kafka/server/DelayedRemoteListOffsets.scala:
```diff
@@ -0,0 +1,174 @@
+              case Right(Some(found)) =>
+                var partitionResponse = buildErrorResponse(Errors.NONE, partition.partition())
+                if (status.lastFetchableOffset.isDefined && found.offset >= status.lastFetchableOffset.get) {
+                  if (status.maybeOffsetsError.isDefined) {
+                    val error = if (version >= 5) Errors.forException(status.maybeOffsetsError.get) else Errors.LEADER_NOT_AVAILABLE
+                    partitionResponse.setErrorCode(error.code())
+                  }
```
It's weird there's no `else` case when lastFetchableOffset <= found offset. From the original logic, we  should still return error, right?

kamalcph replied:
The original logic returns `empty` when  lastFetchableOffset <= found and maybeOffsetsError is undefined. If the result is empty, then we return (no-error, -1, -1L) from KafkaApis.scala:

1. https://sourcegraph.com/github.com/apache/kafka@049b7cde4c2f75a6af164d7308318e29878a2eca/-/blob/core/src/main/scala/kafka/cluster/Partition.scala?L1617
2. https://sourcegraph.com/github.com/apache/kafka@trunk/-/blob/core/src/main/scala/kafka/server/KafkaApis.scala?L1218

showuon commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1447,11 +1466,119 @@ class ReplicaManager(val config: KafkaConfig,
+    }
+
+    if (delayedRemoteListOffsetsRequired(statusByPartition)) {
+      val timeout = config.remoteLogManagerConfig.remoteListOffsetsRequestTimeoutMs()
```
So we will include client timeout value in the follow-up PR?

kamalcph replied:
yes, in #17112

showuon reviewed (commented):
Had another look, left some comments. Thanks for the PR.

kamalcph committed:
Addressed the review comments

- Started the key-value pair from 0 to match with the offset number. (k0, v0) matches with the offset-0, this improves the test readability.

kamalcph committed:
Run the ListOffsetsTest only for the Kraft mode.

kamalcph committed:
apply spotlessbugs java check

kamalcph commented:
@showuon 
Addressed your review comments. PTAL. Thanks!

showuon reviewed (approved):
LGTM! Thanks for the PR!

chia7712 commented:
@kamalcph Could you please fix the conflicts?

kamalcph committed:
fix the conflicts

satishd reviewed (commented):
Thanks @kamalcph for the PR. Overall LGTM, once the conflicts are solved.

chia7712 commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -620,6 +631,23 @@ private Optional<Integer> maybeLeaderEpoch(int leaderEpoch) {
+                    delayedRemoteListOffsetsPurgatory.checkAndComplete(key);
+                })
+        );
+        return new AsyncOffsetReadFutureHolder<>(jobFuture, taskFuture);
```
Pardon me, why we need two futures here? Is `CompletableFuture.supplyAsync` unsuitable to this case? for example:
```java
        CompletableFuture<Optional<FileRecords.TimestampAndOffset>> taskFuture = CompletableFuture.supplyAsync(() -> {
            try {
                // If it is not found in remote storage, then search in the local storage starting with local log start offset.
                Optional<FileRecords.TimestampAndOffset> rval = findOffsetByTimestamp(topicPartition, timestamp, startingOffset, leaderEpochCache);
                if (rval.isPresent()) return rval;
                return OptionConverters.toJava(searchLocalLog.get());
            } catch (Exception e) {
                // NOTE: All the exceptions from the secondary storage are catched instead of only the KafkaException.
                LOGGER.error("Error occurred while reading the remote log offset for {}", topicPartition, e);
                throw new RuntimeException(e);
            }
        }, remoteStorageReaderThreadPool);
```

kamalcph replied:
Thanks for the review! 

The reason for maintaining 2 futures: jobFuture and taskFuture. They are required to trigger the delayed operation completion (delayedRemoteListOffsetsPurgatory#checkAndComplete(key)) in the same remote-log-reader thread after the `RemoteLogManager#findOffsetByTimestamp ++ searchLocalLog` operation completes. 

In DelayedRemoteListOffsets purgatory, we return the result when all the partitions results are received. Then, the delayedOperation gets completed.

We have ActionQueue to complete the pending actions but the LIST_OFFSETS request can be served by any replica (least loaded node). If the node serving the request doesn't have leadership for any of the partitions, then the result might not be complete.

chia7712 replied:
Thanks for the explanation, but "trigger the delayed operation completion in the same thread" seems to work by `CompletableFuture.supplyAsync`, right?
```java
        CompletableFuture<Either<Exception, Option<FileRecords.TimestampAndOffset>>> taskFuture = CompletableFuture.supplyAsync(() -> {
            Either<Exception, Option<FileRecords.TimestampAndOffset>> result;
            try {
                // If it is not found in remote storage, then search in the local storage starting with local log start offset.
                Option<FileRecords.TimestampAndOffset> timestampAndOffsetOpt =
                        OptionConverters.toScala(findOffsetByTimestamp(topicPartition, timestamp, startingOffset, leaderEpochCache))
                                .orElse(searchLocalLog::get);
                result = Right.apply(timestampAndOffsetOpt);
            } catch (Exception e) {
                // NOTE: All the exceptions from the secondary storage are catched instead of only the KafkaException.
                LOGGER.error("Error occurred while reading the remote log offset for {}", topicPartition, e);
                result = Left.apply(e);
            } finally {
                TopicPartitionOperationKey key = new TopicPartitionOperationKey(topicPartition.topic(), topicPartition.partition());
                delayedRemoteListOffsetsPurgatory.checkAndComplete(key);
            }
            return result;
        }, remoteStorageReaderThreadPool);
```

I notice there is another similar pattern `DelayedRemoteFetch`, so it is OK to keep current design for consistency. However, it would be great to let me known (for my own education) what the side-effect happens if using `CompletableFuture.supplyAsync` :)

kamalcph replied:
This won't work. When `delayedRemoteListOffsetsPurgatory.checkAndComplete(key)` is invoked, it calls `DelayedRemoteListOffset#tryComplete` which checks whether the taskFuture gets completed or not. It is not completed so the delayedOperation won't complete.

chia7712 replied:
> It is not completed so the delayedOperation won't complete.

oh, you are totally right, thanks!!!

chia7712 commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1447,11 +1466,119 @@ class ReplicaManager(val config: KafkaConfig,
+              fetchOnlyFromLeader)
+
+            val status = resultHolder match {
+              case OffsetResultHolder(Some(found), _) =>
```
The data structure gets complicated in this path. If those new structures serve for "remote" only, could you please consider defining a subclass of `TimestampAndOffset` to have data used by remote only?

or please add comments for those cases at least?

kamalcph replied:
This PR is large. I'll address the refactoring in the next PR. Is it fine?

>  please add comments for those cases at least?

Added comments. Let me know if it needs to be improved.

chia7712 commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -1336,34 +1336,39 @@ class UnifiedLog(@volatile var logStartOffset: Long,
           .find(_.maxTimestamp() == maxTimestampSoFar.timestamp)
           .flatMap(batch => batch.offsetOfMaxTimestamp().asScala.map(new TimestampAndOffset(batch.maxTimestamp(), _,
             Optional.of[Integer](batch.partitionLeaderEpoch()).filter(_ >= 0))))
+        OffsetResultHolder(timestampAndOffsetOpt)
```
Pardon me, why `MAX_TIMESTAMP` does not consider the records in the remote storage?

kamalcph replied:
Went over [KIP-734](https://cwiki.apache.org/confluence/display/KAFKA/KIP-734%3A+Improve+AdminClient.listOffsets+to+return+timestamp+and+offset+for+the+record+with+the+largest+timestamp), the purpose of MAX_TIMESTAMP is to get the offset of the record with highest timestamp in the partition:

```
Used to retrieve the offset with the largest timestamp of a partition as message timestamps can be specified client side this may not match the log end offset returned by LatestSpec
```

With remote storage enabled, all the passive segments might be uploaded to remote and removed from local-log. The local-log might contain only one **empty** active segment. We have to handle the `MAX_TIMESTAMP` case for remote storage. Thanks for catching this! Filed KAFKA-17552 to track this issue separately.

kamalcph replied:
I went over #15621 to see how do we handle the MAX_TIMESTAMP case for normal topics.

Now, we maintain the [shallowOffsetOfMaxTimestampSoFar](https://sourcegraph.com/github.com/apache/kafka@trunk/-/blob/storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java?L212) in LogSegment instead of the real-max-timestamp-offset. While uploading the LogSegment to remote, we create the [RemoteLogSegmentMetadata](https://sourcegraph.com/github.com/apache/kafka@trunk/-/blob/storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentMetadata.java) event which holds the metadata information about the segment. 

Even, if we pass the `shallowOffsetOfMaxTimestampSoFar` in the RemoteLogSegmentMetadata event, we have to download the remote-log-segment to find the real `offsetOfMaxTimestampSoFar`. This will increase the load on remote storage assuming that the original intention of the KIP-734 is to `Confirming topic/partition "liveness"` which means the Admin client will repeatedly invoke the list-offsets API for MAX_TIMESTAMP. 

The predominant case to find the "Confirming topic/partition livness" is to query the local-log which will work as expected. For MAX_TIMESTAMP, when enabled with remote storage, the results can go wrong when:

1. Assuming the timestamps are monotonic and all the passive segments are uploaded to remote and deleted from local. The only active segment in local disk is empty (post the log roll), then the max-timestamp will returned as "-1".
2. Assuming the timestamps are non-monotonic, then the (timestamp, offset) returned by the API may not be "TRUE" (max-timestamp, max-timestamp-offset) as it considers the segments only in the local-log.

Should we handle/drop the MAX_TIMESTAMP case for topics enabled with remote storage? This can cause high load:

1. On the RemoteLogMetadataManager, as we have to scan all the uploaded segment events to find the max-timestamp. Then, compare it with the max-timestamp computed from the local-log segments. And, the search should always proceed from remote to local storage.
2. On the RemoteStorageManager, when there exists the MAX_TIMESTAMP record in the remote storage, then we have to download that segment (few bytes) repeatedly to serve the query. 


In [KIP-734](https://cwiki.apache.org/confluence/display/KAFKA/KIP-734%3A+Improve+AdminClient.listOffsets+to+return+timestamp+and+offset+for+the+record+with+the+largest+timestamp), can we make a addendum to say that MAX_TIMESTAMP is not supported for topics enabled with remote storage? Note that when the KIP was proposed, the intention was not to read from the disk:

Snippet from [KIP-734](https://cwiki.apache.org/confluence/display/KAFKA/KIP-734%3A+Improve+AdminClient.listOffsets+to+return+timestamp+and+offset+for+the+record+with+the+largest+timestamp):
```
LogSegments track the highest timestamp and associated offset so we don't have to go to disk to fetch this
```

cc @satishd @showuon

chia7712 replied:
> Even, if we pass the shallowOffsetOfMaxTimestampSoFar in the RemoteLogSegmentMetadata event, we have to download the remote-log-segment to find the real offsetOfMaxTimestampSoFar.

IMHO, we don't need to pass `shallowOffsetOfMaxTimestampSoFar` to `RemoteLogSegmentMetadata` as `RemoteLogSegmentMetadata` has `maxTimestampMs` already, so the basic behavior is listed below (same as you described I'd say)
1. find the max timestamp from local segments
2. query `findOffsetByTimestamp` by max timestamp from step_1
3. compare the timestamp of record from remote to local to pick up correct offset

>  This will increase the load on remote storage assuming that the original intention of the KIP-734 is to Confirming topic/partition "liveness" which means the Admin client will repeatedly invoke the list-offsets API for MAX_TIMESTAMP.

The impl of KIP-734 was wrong because we don't loop all records in all path (because of cost issue). Hence,  we rename the `offsetOfMaxTimestampSoFar` to `shallowOffsetOfMaxTimestampSoFar` based on true story.

> Should we handle/drop the MAX_TIMESTAMP case for topics enabled with remote storage? This can cause high load:

that is a acceptable approach. We can REJECT the MAX_TIMESTAMP request for now as it is rare operation. Or we can make the call an async op too as it needs to iterate all metadata of remote segments.

chia7712 commented on core/src/main/scala/kafka/server/DelayedRemoteListOffsets.scala:
```diff
@@ -0,0 +1,174 @@
+  override def tryComplete(): Boolean = {
+    var completable = true
+    metadata.statusByPartition.forKeyValue { (partition, status) =>
+      if (!status.completed) {
```
Could we check the status of futures instead? for example:
```
  def completable = status.futureHolderOpt.isEmpty || status.futureHolderOpt.get.jobFuture.isDone
```

kamalcph replied:
As mentioned in the other comment, we need the `status.completed` variable as it is `volatile` and accessed by multiple threads for inter-thread visibility.

kamalcph replied:
>   def completable = status.futureHolderOpt.isEmpty || status.futureHolderOpt.get.jobFuture.isDone

This method seems to be thread-safe. We can use this method instead of `completed` variable. But, I find using the variable makes the code clear. I'm open to change this. Will take this refactoring in the next PR.

chia7712 commented on core/src/main/scala/kafka/server/DelayedRemoteListOffsets.scala:
```diff
@@ -0,0 +1,174 @@
+  metadata.statusByPartition.forKeyValue { (topicPartition, status) =>
+    status.completed = status.futureHolderOpt.isEmpty
+    if (status.futureHolderOpt.isDefined) {
+      status.responseOpt = Some(buildErrorResponse(Errors.REQUEST_TIMED_OUT, topicPartition.partition()))
```
Pardon me, why setting `responseOpt` early? If we keep the `responseOpt` be None, we can reuse `responseOpt` to evaluate "completed"

kamalcph replied:
yes, that's true. Assume that one LIST_OFFSETS request wants to query the `offsetForTimestamp` for 10 partitions, those 10 partitions are handled in concurrent fashion, provided that the remote-log-reader threads are available. If any one thread completes, then it marks that partition status as completed and checks for the statuses of all the other partitions. 

The variable `completed` is accessed by multiple remote-log-reader threads so marked it as volatile and used it for computation instead of `responseOpt`.

chia7712 reviewed (commented):
@kamalcph thanks for this patch and sorry for late reviews. a couple of comments are left. PTAL

kamalcph committed:
added comments.

chia7712 reviewed (approved):
LGTM

satishd reviewed (approved):
Thanks @kamalcph  for addressing the review comments. LGTM.

kamalcph commented:
JDK11 test failures are unrelated to this PR, the tests were timed out.

```
Found 2 flaky test failures:
FLAKY ⚠️  MetricsDuringTopicCreationDeletionTest > "testMetricsDuringTopicCreateDelete(String).quorum=zk"
FLAKY ⚠️  OffsetsApiIntegrationTest > testAlterSinkConnectorOffsetsDifferentKafkaClusterTargeted()
Read env GRADLE_EXIT_CODE: 124
Read env THREAD_DUMP_URL: https://github.com/apache/kafka/actions/runs/10854248301/artifacts/1932678337
Gradle command timed out. These are partial results!
25602 tests cases run in 6h14m4s. 23379 PASSED ✅, 0 FAILED ❌, 2 FLAKY ⚠️ , 18 SKIPPED 🙈, and 0 errors.
Failing this step because the tests timed out. Thread dumps were taken and archived here: https://github.com/apache/kafka/actions/runs/10854248301/artifacts/1932678337
Error: Process completed with exit code 1.
```

chia7712 commented:
@kamalcph thanks for checking the failed tests. They pass on my local. will merge this PR

```
./gradlew cleanTest :tools:test --tests MetadataQuorumCommandTest.testDescribeQuorumReplicationSuccessful :core:test --tests ListOffsetsIntegrationTest.testThreeNonCompressedRecordsInSeparateBatch --tests ConsumerBounceTest.testConsumptionWithBrokerFailures
```

chia7712 commented:
@kamalcph any updates? or just trigger QA again?

kamalcph commented:
I rebased the based against trunk to retrigger the tests again.

chia7712 commented:
> I rebased the branch against trunk to retrigger the tests again.

Got it

kamalcph commented:
Thank you all for the reviews!

mumrah commented:
@kamalcph, the test introduced in this PR is flaky on trunk https://ge.apache.org/scans/tests?search.names=CI%20workflow&search.rootProjectNames=kafka&search.tags=github%2Ctrunk&search.tasks=test&search.timeZoneId=America%2FNew_York&search.values=CI&tests.container=kafka.log.remote.RemoteLogOffsetReaderTest&tests.sortField=FLAKY

Can you take a look? I have filed KAFKA-17559 in the mean time.

chia7712 commented:
@mumrah Sorry for that flaky. I will take a look!

kamalcph commented:
Opened #17214 to fix the flaky test. PTAL.

junrao commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -1263,7 +1263,7 @@ class UnifiedLog(@volatile var logStartOffset: Long,
    */
   @nowarn("cat=deprecation")
-  def fetchOffsetByTimestamp(targetTimestamp: Long, remoteLogManager: Option[RemoteLogManager] = None): Option[TimestampAndOffset] = {
+  def fetchOffsetByTimestamp(targetTimestamp: Long, remoteLogManager: Option[RemoteLogManager] = None): OffsetResultHolder = {
```
Could we change the description of the return value accordingly?

kamalcph replied:
Addressed this in #17487.

junrao commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -263,6 +270,10 @@ public RemoteLogManager(RemoteLogManagerConfig rlmConfig,
     }
 
+    public void setDelayedOperationPurgatory(DelayedOperationPurgatory<DelayedRemoteListOffsets> delayedRemoteListOffsetsPurgatory) {
+        this.delayedRemoteListOffsetsPurgatory = delayedRemoteListOffsetsPurgatory;
```
delayedRemoteListOffsetsPurgatory is written and read by different threads. Does it need to be volatile?

kamalcph replied:
Addressed this in https://github.com/apache/kafka/pull/17487.

kamalcph replied:
Other purgatories are also accessed by multiple threads but they don't have the `volatile`. So, followed the same approach. Let me know whether it is required.

junrao replied:
You are talking about the purgatories in `ReplicaManager`? They are set during the creation of ReplicaManager. Here, `delayedRemoteListOffsetsPurgatory` is not set during the creation of `RemoteLogManager`.

kamalcph replied:
Added the `volatile` to the delayedRemoteListOffsetsPurgatory in RemoteLogManager. My understanding was that we instantiate the [dataPlaneRequestProcessor](https://sourcegraph.com/github.com/apache/kafka/-/blob/core/src/main/scala/kafka/server/KafkaServer.scala?L604-608) after calling the [ReplicaManager#startup](https://sourcegraph.com/github.com/apache/kafka/-/blob/core/src/main/scala/kafka/server/ReplicaManager.scala?L418), so there won't be an issue. It is good to be on the safer side.

junrao commented on core/src/main/scala/kafka/server/DelayedRemoteListOffsets.scala:
```diff
@@ -0,0 +1,174 @@
+  }
+}
+
+class DelayedRemoteListOffsets(delayMs: Long,
```
`ReplicaManager` calls `completeDelayedFetchOrProduceRequests` when a replica is removed from the broker or becomes a follower to wake up pending produce/fetch request early. Should we do the same for pending remoteListOffset requests?

kamalcph replied:
Addressed this in https://github.com/apache/kafka/pull/17487. 

Retained the same method name. Shall we change the method name to `completeDelayedFetchOrProduceOrRemoteListOffsetsRequests`?

[ReplicaManager.completeDelayedFetchOrProduceRequests](https://sourcegraph.com/github.com/apache/kafka/-/blob/core/src/main/scala/kafka/server/ReplicaManager.scala?L465-467)

junrao replied:
Perhaps naming it `completeDelayedOperationsWhenNotPartitionLeader` ?

kamalcph replied:
Renamed the method.

junrao commented on core/src/main/scala/kafka/server/DelayedRemoteListOffsets.scala:
```diff
@@ -0,0 +1,174 @@
+import scala.collection.{Map, mutable}
+import scala.jdk.CollectionConverters._
+
+case class ListOffsetsPartitionStatus(var responseOpt: Option[ListOffsetsPartitionResponse] = None,
```
`responseOpt` can be written and read by different threads. Should it be volatile?

kamalcph replied:
Addressed this in https://github.com/apache/kafka/pull/17487.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1447,11 +1466,123 @@ class ReplicaManager(val config: KafkaConfig,
+      // create a list of (topic, partition) pairs to use as keys for this delayed remote list offsets operation
+      val listOffsetsRequestKeys = statusByPartition.keys.map(TopicPartitionOperationKey(_)).toSeq
+      // try to complete the request immediately, otherwise put it into the purgatory
+      delayedRemoteListOffsetsPurgatory.tryCompleteElseWatch(delayedRemoteListOffsets, listOffsetsRequestKeys)
```
`listOffsetsRequestKeys` is a bit weird. It is based on a topicPartition and the purgatory is checked on that key every time a remote listOffset task completes. However, the completion of such a task has no impact on other pending listOffset requests on the same partition.

The only reason we need the purgatory is really just for the expiration logic after the timeout if we chain all the futures together. Perhaps, using the pattern of DelayedFuturePurgatory is more intuitive?

kamalcph replied:
Went through the `DelayedFuturePurgatory` and understood the changes required. This is a big change, so we can take this separately: 

1. DelayedFuturePurgatory does not emit any request expiration metrics.
2. When a replica for a partition moves away, then we cannot complete the request for that partition as the `watchKey` is unique for each request.

junrao replied:
1. Where is the request expiration metric emitted now?
2. Good point. We could have a customized DelayedFuturePurgatory that also adds a delayed operation key per partition. But they are triggered for completion check when the replica is no longer the leader.

kamalcph replied:
1. The expiration metrics are being emitted by the individual purgatory. (eg) [DelayedRemoteListOffsetsMetrics#recordExpiration](https://sourcegraph.com/github.com/apache/kafka/-/blob/core/src/main/scala/kafka/server/DelayedRemoteListOffsets.scala?L76)
2. Agree, this will improve the performance. Will take the custom DelayedFuturePurgatory changes separately as it is a big change.

kamalcph replied:
Filed KAFKA-17797 to track this change.

junrao commented on core/src/main/scala/kafka/server/DelayedRemoteListOffsets.scala:
```diff
@@ -0,0 +1,174 @@
+  }
+}
+
+case class ListOffsetsMetadata(statusByPartition: mutable.Map[TopicPartition, ListOffsetsPartitionStatus]) {
```
Hmm, why do we need this wrapper class? Could we just use the `Map` directly?

kamalcph replied:
Addressed this in https://github.com/apache/kafka/pull/17487.

junrao reviewed (commented):
@kamalcph : Thanks for the PR. Sorry for the late review. Added a few more comments.

kamalcph commented:
Thanks @junrao for the review comments! I will follow-up on them.
